### ISSUE TITLE
emplace_hint missing in gcc 4.6

### ISSUE THREAD

**Issue by tjhei** [geodynamics/aspect#1260]
I am getting
```
../source/particle/world.cc:525:15: error: ‘class std::multimap<std::pair<int, int>, aspect::Particle::Particle<2>, std::less<std::pair<int, int> >, std::allocator<std::pair<const std::pair<int, int>, aspect::Particle::Particle<2> > > >’ has no member named ‘emplace_hint’
```
with gcc 4.6.3 (CXX11=true, CXX14=false).

What should we do: Add a configure test? Only enable if we have c++14? Any other suggestions?

**Comment by bangerth** [comment]
Ah, boo. @gassmoeller -- any idea?

**Comment by gassmoeller** [comment]
Yeah my reaction as well :smile: ... seems like even gcc 4.7 does not have full support for the emplace functions. I would argue to change it to C++14. This should not give a performance hit, because with C++11 we have the move constructor for particles, and that should make the two versions more or less equivalent anyway (one call to the normal constructor of Particle, one to the move constructor / maybe two move constructor calls for the `insert` version, I have not checked). Then we can use the real benefit of `emplace_hint` with C++14, namely the piecewise construction of the members of the pair (no more move constructor calls). I can prepare a PR for that later, but feel free to go ahead and change the requirement @tjhei.

**Comment by tjesser-ucdavis-edu** [comment]
Are gcc 4.6 and 4.7 versions required for compatibility somewhere? I noticed that 4.6 hasn't had a release since 2013, and 4.7 hasn't had a release since 2014. https://gcc.gnu.org/develop.html#timeline

I ask this assuming that some later version of gcc has implemented C++11 correctly.

**Comment by gassmoeller** [comment]
At least 4.8.4 is fine (I think all of the 4.8 series is complete in C++11). We should try to keep compatibility to older compilers though, since many group-managed clusters tend to use very old compilers (stable CentOS or RHEL releases, and no admins to update compilers). One of the clusters I know (and my current work desktop :smile:) use gcc 4.4.7.

**Comment by gassmoeller** [comment]
I just remembered: I think deal.II plans to require C++11 some time early next year. When that happens we can clearly follow with ASPECT, but I would want to wait until then (e.g. the first release of ASPECT after deal.II has made the first release with that requirement).

**Comment by bangerth** [comment]
@tjesser-ucdavis-edu -- for the moment, ASPECT does not require C++11, nor any particularly new version of GCC. But we have a few places where if you happen to have a newer compiler, we use more efficient code paths. There is, however, always a fall-back path in case you don't.

As @gassmoeller says, we will probably require C++11 for deal.II sometime next year. I'm sure we will make another ASPECT release before that, but eventually ASPECT will also require that.

**Comment by bangerth** [comment]
I think this issue has been fixed by #1265.
